pkg/client/mutator_test.go

```diff
@@ -0,0 +1,336 @@
+	assert.Nil(t, err)
+
+	testClient := &Client{
+		RESTClient: testRestClient,
+		Config:     testConfig,
+	}
+
+	testClientWithAgent, err := testClient.WithAgent("testagent")
+	assert.Nil(t, err)
+
+	testRestClientWithAgent := testClientWithAgent.RESTClient.(*rest.RESTClient)
+
+	_, err = testRestClientWithAgent.Client.Transport.RoundTrip(&http.Request{})
+	assert.Nil(t, err)
+
+	// with invalid config
+	testClient = &Client{
+		Config: rest.Config{},
+	}
+
+	testClientWithAgent, err = testClient.WithAgent("testagent")
+	assert.NotNil(t, err)
+	assert.Nil(t, testClientWithAgent)
+}
+
+func TestWithImpersonate(t *testing.T) {
+	testUserAgent := "testagent"
+	testConfig := rest.Config{
+		UserAgent: "defaultuseragent",
+		ContentConfig: rest.ContentConfig{
+			NegotiatedSerializer: serializer.WithoutConversionCodecFactory{},
+		},
+		Transport: userAgentRoundTripper{t: t, expectedUserAgent: testUserAgent},
+	}
+
+	testRestClient, err := rest.UnversionedRESTClientFor(&testConfig)
+	assert.Nil(t, err)
+
+	testClient := &Client{
+		RESTClient: testRestClient,
+		Config:     testConfig,
+	}
+
+	testClientWithAgent, err := testClient.WithAgent("testagent")
+	assert.Nil(t, err)
+
+	testRestClientWithAgent := testClientWithAgent.RESTClient.(*rest.RESTClient)
+
+	_, err = testRestClientWithAgent.Client.Transport.RoundTrip(&http.Request{})
+	assert.Nil(t, err)
+
+	// with invalid config
+	testClient = &Client{
+		Config: rest.Config{},
+	}
+
+	testClientWithAgent, err = testClient.WithAgent("testagent")
+	assert.NotNil(t, err)
+	assert.Nil(t, testClientWithAgent)
+}
+
+func TestNewSharedClientFactoryWithAgent(t *testing.T) {
+	mockSharedClientFactory := NewMockSharedClientFactory(gomock.NewController(t))
+
+	testUserAgent := "testagent"
+	testConfig := rest.Config{
+		UserAgent: "defaultuseragent",
+		ContentConfig: rest.ContentConfig{
+			NegotiatedSerializer: serializer.WithoutConversionCodecFactory{},
+		},
+		Transport: userAgentRoundTripper{t: t, expectedUserAgent: testUserAgent},
+	}
+
+	testRestClient, err := rest.UnversionedRESTClientFor(&testConfig)
+	assert.Nil(t, err)
+	testClient := &Client{
+		RESTClient: testRestClient,
+		Config:     testConfig,
+	}
+	mockSharedClientFactory.EXPECT().ForKind(gomock.Any()).DoAndReturn(func(gvk schema.GroupVersionKind) (*Client, error) {
+		return testClient, nil
+	}).AnyTimes()
+	mockSharedClientFactory.EXPECT().ForResource(gomock.Any(), gomock.Any()).DoAndReturn(func(gvr schema.GroupVersionResource, namespaced bool) (*Client, error) {
+		return testClient, nil
+	}).AnyTimes()
+	mockSharedClientFactory.EXPECT().ForResourceKind(gomock.Any(), gomock.Any(), gomock.Any()).DoAndReturn(func(gvr schema.GroupVersionResource, kind string, namespaced bool) *Client {
+		return testClient
+	}).AnyTimes()
+	testSharedClientFactoryWithAgent := NewSharedClientFactoryWithAgent("testagent", mockSharedClientFactory)
+
+	// for kind
+	clientWithAgent, err := testSharedClientFactoryWithAgent.ForKind(schema.GroupVersionKind{})
```

Comment:
They shouldn't be tested independently from with `WithAgent` or `WithImpersonation` since a user can not use them unless they call `WithAgent` or `WithImpersonation`. Furthermore, if we were to test them with the private structure the test would look exactly the same. We would call the functions `ForKind`, `ForResource`, and `ForResourceKind` and verify there was a mutation made i.e. user agent was set. The only thing that would change is the name of the test function.

Reply:
That's a good point on them being a part of the same test. We can consider that part resolved.

However, I do think that it would be a good idea to have a component of the round tripper which makes an assertion that the impersonation header fields were set properly. We have this for `userAgent` - the previous implementation uses this fake `userAgentRoundTripper` to assert the user-agent header was set properly. It would be a good idea to do this for the  impersonation header too.

Reply:
> it would be a good idea to do this for the impersonation header too.
This was already being done in the test.

And I added a separate test for `ForKind, ForResource, and ForResourceKind` for brevity.